### ISSUE TITLE
Please don't impose Werror on downstream users

### ISSUE THREAD

**Issue by shlevy** [input-output-hk/cardano-base#372]
`Werror` is great for development and CI, but is not appropriate for users of your library. New compiler versions frequently bring out new warnings or refine their heuristics. When this breaks the build, all downstream can reasonably do is blindly disable Werror anyway.

Fresh project using `plutus-tx` (and thus, indirectly, `cardano-crypto-class`) fails with GHC 9.4.6 with:

```
[ 6 of 48] Compiling Cardano.Crypto.Hash.Class ( src/Cardano/Crypto/Hash/Class.hs, dist/build/Cardano/Crypto/Hash/Clas>

src/Cardano/Crypto/Hash/Class.hs:173:38: error: [-Wdeprecations, -Werror=deprecations]
    In the use of ‘serializeEncoding'’
    (imported from Cardano.Binary, but defined in cardano-binary-1.6.0.0-9zN4iiyMBUGJoTfCIR8U3A:Cardano.Binary.Seriali>
    Deprecated: "Use `serialize'` instead, since `Encoding` has `ToCBOR` instance"
    |
173 | hashWithSerialiser toEnc = hashWith (serializeEncoding' . toEnc)
    |                                      ^^^^^^^^^^^^^^^^^^
```

**Comment by shlevy** [comment]
I will work around this for now with the ironically named `development` flag

**Comment by michaelpj** [comment]
Yes, please don't do this. The package is unusable from CHaP at the moment because of this. In future we will at least check, but `-Werror` is a plague.

**Comment by erikd** [comment]
In `node` -Werror` is not used in any of the `<package.cabal>` files but in the `cabal.project` file we have:

```
package cardano-api
  ghc-options: -Werror

package cardano-cli
  ghc-options: -Werror

package cardano-git-rev
  ghc-options: -Werror
.....
```

**Comment by angerman** [comment]
You could even do
```
package *
   ghc-options: -Werror
```

though that might go a bit overboard. I agree that `-Werror` should _not_ be in `<pkg>.cabal` files, but should be end-user driven in the `cabal.project` file. One is opt-in (cabal.project), and the other is not even opt-out. Or can you set `-Wno-error`?

**Comment by lehins** [comment]
Actually that's the proper flag in `project.cabal`:
```
program-options
  ghc-options: -Werror
```

Because i believe this will also turn on `-Werror` for dependencies:
```
package *
   ghc-options: -Werror
```

>  I think it's best to just turn on -Werror in CI, which is usually pretty easy.

I think it is best to turn it on in the `project.cabal` file for all packages in the project, because it will prevent new contributors from getting confused with: "It builds for me locally why is it failing on CI?"

Also some people prefer to have `-Werror` enabled locally, because errors force you to recompile, while warnings are not triggered on recompilation and can easily sneak in into a PR. This will of course trigger the CI failure, but some people prefer to get those warnings the first time around. I am not one of those people, I am just conveying opinions of others, and I'd like to respect opinions of others :slightly_smiling_face: 

Anyone willing to review this PR with a fix: https://github.com/input-output-hk/cardano-base/pull/376

**Comment by michaelpj** [comment]
> I think it is best to turn it on in the project.cabal file for all packages in the project, because it will prevent new contributors from getting confused with: "It builds for me locally why is it failing on CI?"

Just as an empirical data point, this has never happened on `plutus` and we do get external contributors from time to time.

Anyway, you can do what you like with your own codebase, but it really is very unhelpful if it gets into cabal files in CHaP, so I'm happy we're avoiding that.

**Comment by lehins** [comment]
> it really is very unhelpful if it gets into cabal files in CHaP,

Just to correct myself and remove any ambiguity. I made a slight mistake, which I do all the time. I said it is better to add `-Werror` to `project.cabal`, where I meant say `cabal.project`. (Trips me up every time) :slightly_smiling_face: This does ensure that it does not end up in CHaPs, cause that part was definitely bad.

> Just as an empirical data point, this has never happened on plutus 

It doesn't mean it didn't happen, maybe just nobody admitted to it :stuck_out_tongue_winking_eye:

**Comment by coot** [comment]
I also agree with @michaelpj, we've been using `-Werror` just in CI and it has been working quite well for us.

> Or can you set -Wno-error?
@angerman I had some luck with setting `-Wwarn`.
